memtable.cc

```diff
@@ -702,15 +702,39 @@ memtable::make_flat_reader(schema_ptr s,
         rd.upgrade_schema(s);
         return rd;
     } else {
-        auto res = make_flat_mutation_reader<scanning_reader>(std::move(s), shared_from_this(), std::move(permit), range, slice, pc, fwd_mr);
-        if (fwd == streamed_mutation::forwarding::yes) {
-            return make_forwardable(std::move(res));
-        } else {
-            return res;
-        }
+        return make_flat_mutation_reader<scanning_reader>(std::move(s), shared_from_this(), std::move(permit), range, slice, pc, fwd_mr);
     }
 }
 
+flat_mutation_reader
+memtable::make_flat_reader(schema_ptr s, reader_permit permit, const dht::partition_range& range, const query::partition_slice& query_slice,
+        const io_priority_class& pc, tracing::trace_state_ptr trace_state_ptr, streamed_mutation::forwarding fwd, mutation_reader::forwarding fwd_mr) {
+    // When the memtable is flushed while a read is ongoing, an sstable reader is
+    // created to replace the memtable reader mid-air. This sstable reader will
```

Comment:
Note that this happens only for multi-partition reads. The switch happens when crossing partition boundaries.
For single-partition reads the partition being read is a snapshot that stays in memory even after flush (if there's an ongoing read from the snapshot).

Reply:
We can use the non-reversed slice uniformly, it is the safer option.

Reply:
I agree, I was just referring to the comment which may confuse a new person who comes and reads this (they may think that the delegating to sstable reader happens for single-partition queries too).

Reply:
I see, I can clarify that.